zashi-pipeline.py

```diff
@@ -0,0 +1,217 @@
+        for (workspace_id, repos) in workspaces.items()
+        if len(repos) > 0
+    ])
+
+    print('Rendering deployment pipeline')
+
+    # Ensure that the tracked issues all exist in the graph. This is a no-op for
+    # issues that are already present.
+    start_at = set([issue for issue in tracked_issues])
+    for i in start_at:
+        dg.add_node(i)
+
+    # Replace the graph with the subgraph that only includes the tracked
+    # issues and their descendants.
+    descendants = [nx.descendants(dg, n) for n in start_at]
+    dg = nx.subgraph(dg, start_at.union(*descendants))
+
+    # Fetch the issues within the graph.
+    mapping = github.download_issues(gapi, dg.nodes, repos)
+
+    # Relabel the graph
+    dg = nx.relabel_nodes(dg, mapping)
+
+    # Filter out unknown issues
+    unknown = [n for n in dg if n.repo_id not in repos]
+    if len(unknown) > 0:
+        dg.remove_nodes_from(unknown)
+
+    # Apply property annotations
+    for (source, sink) in dg.edges:
+        attrs = dg.edges[source, sink]
+        attrs['is_open'] = 0 if source.state == 'closed' else 1
+
+    # Render the HTML version!
+    html_header = '''<!DOCTYPE html>
+<html>
+  <head>
+    <title>Zashi Pipeline</title>
+
+    <style>
+      body {
+        color: #1f2328;
+        font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Noto Sans",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji";
+        font-size: 14px;
+        line-height: 1.5;
+        word-wrap: break-word;
+      }
+      a {
+        color: #0969da;
+      }
+      table {
+        border-collapse: collapse;
+        width: 100%;
+        width: max-content;
+        max-width: 100%;
+        overflow: auto;
+      }
+      table tr {
+        border-top: 1px solid #d1d9e0b3;
+      }
+      table th, table td {
+        border: 1px solid #d1d9e0b3;
+        padding: 6px 13px;
+        text-align: center;
+      }
+
+      @media (prefers-color-scheme: dark) {
+        body {
+          background-color: #121212;
+          color: #f0f6fc;
+        }
+        a {
+          color: #4493f8;
+        }
+        table tr {
+            border-top: 1px solid #3d444db3;
+        }
+        table th, table td {
+            border: 1px solid #3d444db3;
+        }
+      }
+    </style>
+  </head>
+  <body>
+    <h1>Zashi Pipeline</h1>
+    <p>🐞 = bug, 💡 = feature, ✅ = implemented / released, 🛑 = unfinished, 📥 = untracked</p>
+    <table>
+      <thead>
+        <tr>
+          <th>Type</th>
+          <th>Issue</th>
+          <th>Rust crate</th>
+          <th>Android SDK</th>
+          <th>Swift SDK</th>
+          <th>Zashi Android</th>
+          <th>Zashi iOS</th>
+        </tr>
+      </thead>
+      <tbody>
+'''
+    html_footer = '''
+      </tbody>
+    </table>
+  </body>
+</html>
+'''
+    with open('public/zcash-change-deployment-pipeline.html', 'w') as f:
+        f.write(html_header)
+
+        for issue in tracked_issues.values():
```

Comment:
Current limitation: we don't filter out rows where everything is complete. That is intentional for now while we start filling out the DAG with the necessary data, but eventually we'll want to hide rows that are no longer relevant.

Reply:
I think it would be useful to keep either rows that have only been completed recently, or a limited number of rows that have been completed.

Reply:
When an issue was closed is available in the `closedAt` attribute of an issue, documented here: https://docs.github.com/en/graphql/reference/objects#issue